content/blog/2019-09-25-manually-managed-memory.md

```diff
@@ -0,0 +1,303 @@
++++
+title = "Manually Managed Memory in Bril"
+extra.author = "Drew Zagieboylo & Ryan Doenges"
+extra.bio = """
+  [Drew Zagieboylo](https://www.cs.cornell.edu/~dzag/) is a 3rd year PhD student researching Security, HW design and Programming Languages. He enjoys rock climbing and gaming in his free time.
+  [Ryan Doenges](http://ryandoeng.es/) is a 3rd year PhD student studying programming languages for networking.
+"""
++++
+
+### Pointers and Heap Allocated Memory
+
+Our goal was to add *pointer types* to Bril. _Pointers_ represent references to
+manually managed read/write memory cells which can persist outside of function
+scope. Furthermore we support C-style arrays such that pointer arithmetic
+instructions can be used to index into allocated memory regions. Lastly, we
+wished to ensure that value typechecking was still supportable for our new
+instructions (however we did not implement a typechecker). Our pointer types are
+meant only for value checking (i.e. every pointer type totally specifies the
+type of its contents); they do not include bounds or alias information to
+prevent memory safety bugs.
+
+### Design Overview
+
+We added manually managed memory and typed pointers to Bril while keeping the
+layout of data hidden from Bril programs. The API for working with these
+heap-allocated pointers was inspired by [LLVM's manually allocated stack pointer
+API](https://llvm.org/docs/LangRef.html#memory-access-and-addressing-operations).
+Pointer types include the data type to which they refer: no `void*` magic
+allowed. Furthermore, the representation of data in memory is abstract. In
+a type system that just consists of integers, booleans, and pointers this is not
+too strong a statement, but something like LLVM's `getelementptr` would allow
+structs to be added while still hiding type sizes from Bril programs.
+This depends on the fact that, unlike in C, you cannot do bytewise arithmetic on
+Bril pointers to determine the size of things in memory or extract the value of
+a pointer as an integer address.
+
+This design leaves the Bril interpreter/compiler complete freedom to choose how
+data, including pointers, are represented and allocated; this separates the high
+level computaional usage of pointers and the low level (and likely
+platform-dependent) implementation details while still providing usable manually
+managed memory to programs.
+
+#### Pointer Syntax & Representation
+
+We expanded the Bril type syntax with ```ptr<TYPE>```, which denotes a pointer
+to a value in memory of type `TYPE`.
```

Comment:
Maybe it would be nice to clarify how this looks in JSON vs. the text representation, for example? I notice in your implementation that the JSON type is `{"ptr": TYPE}`, which is of course the sensible thing to do.